dm/pkg/shardddl/optimism/lock.go

```diff
@@ -871,3 +714,419 @@ func GetColumnName(lockID, ddl string, tp ast.AlterTableType) (string, error) {
 	}
 	return "", nil
 }
+
+// checkAddDropColumn check for ALTER TABLE ADD/DROP COLUMN statement
+// FOR ADD COLUMN, check whether add column with a different field or add a droped column
+// FOR DROP COLUMN, return the droped column.
+func (l *Lock) checkAddDropColumn(source, schema, table string, ddl string, prevTable, postTable schemacmp.Table) (string, error) {
+	currTable := l.tables[source][schema][table]
+	defer func() {
+		l.tables[source][schema][table] = currTable
+	}()
+
+	l.tables[source][schema][table] = prevTable
+	oldJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	l.tables[source][schema][table] = postTable
+	newJoined, err := l.joinNormalTables()
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	cmp, err := oldJoined.Compare(newJoined)
+	if err != nil {
+		// nolint:nilerr
+		return "", nil
+	}
+
+	if cmp <= 0 {
+		if col, err2 := AddDifferentFieldLenColumns(l.ID, ddl, oldJoined, newJoined); err2 != nil {
+			// check for add column with a larger field len
+			return "", err2
+		} else if _, err2 = AddDifferentFieldLenColumns(l.ID, ddl, postTable, newJoined); err2 != nil {
+			// check for add column with a smaller field len
+			return "", err2
+		} else if len(col) > 0 && l.IsDroppedColumn(source, schema, table, col) {
+			return "", terror.ErrShardDDLOptimismTrySyncFail.Generate(l.ID, fmt.Sprintf("add column %s that wasn't fully dropped in downstream. ddl: %s", col, ddl))
+		}
+	}
+
+	if cmp >= 0 {
+		if col, err2 := GetColumnName(l.ID, ddl, ast.AlterTableDropColumn); err2 != nil {
+			return "", err2
+		} else if len(col) > 0 {
+			return col, nil
+		}
+	}
+	return "", nil
+}
+
+// TrySyncForOneDDL try sync for a DDL operation.
+// e.g. `ALTER TABLE ADD COLUMN a, RENAME b TO c, DROP COLUMN d' will call this func three times.
+// return whether joined table is changed and whether there is a conflict.
+func (l *Lock) TrySyncForOneDDL(source, schema, table string, prevTable, postTable schemacmp.Table) (schemaChanged bool, conflictStage ConflictStage) {
```

Comment:
perfer rename to `trySyncForOneDDL `, because no outer caller and we not use lock in this func